pkg/clustermesh/common/namespacewatcher/watcher.go

```diff
@@ -0,0 +1,480 @@
+	if nw.isNamespaceAnnotated(namespace) {
+		return nw.isNamespaceGlobalByAnnotation(namespace)
+	}
+
+	// For non-annotated namespaces when filtering is active, use the default behavior
+	return nw.config.DefaultGlobalNamespace
+}
+
+func (nw *namespaceWatcher) IsFilteringActive() bool {
+	nw.mu.RLock()
+	defer nw.mu.RUnlock()
+	return nw.isFilteringActive()
+}
+
+func (nw *namespaceWatcher) GetGlobalNamespaces() sets.Set[string] {
+	nw.mu.RLock()
+	defer nw.mu.RUnlock()
+
+	// If namespace-based filtering is not active, return empty set to indicate all namespaces are global
+	if !nw.isFilteringActive() {
+		return sets.New[string]()
+	}
+
+	// When filtering is active, collect all global namespaces from the resource store
+	result := sets.New[string]()
+
+	ctx := context.Background()
+	store, err := nw.namespaceResource.Store(ctx)
+	if err != nil {
+		return result
+	}
+
+	// List all namespaces and check which ones are global
+	allNamespaces := store.List()
+	for _, ns := range allNamespaces {
+		annotationValue, hasAnnotation := annotation.Get(ns, annotation.GlobalNamespace)
+		if hasAnnotation {
+			if annotationValue == "true" {
+				result.Insert(ns.Name)
+			}
+		} else if nw.config.DefaultGlobalNamespace {
+			// Non-annotated namespaces are global if the default is true
+			result.Insert(ns.Name)
+		}
+	}
+
+	return result
+}
+
+func (nw *namespaceWatcher) IsGlobalService(obj interface {
+	GetAnnotations() map[string]string
+	GetNamespace() string
+}) bool {
+	// First check if service has the global annotation
+	if !annotation.GetAnnotationIncludeExternal(obj) {
+		return false
+	}
+
+	// If namespace filtering is active, also check if service is in a global namespace
+	if nw.IsFilteringActive() {
+		if !nw.IsGlobalNamespace(obj.GetNamespace()) {
+			// Service is marked as global but not in a global namespace
+			return false
+		}
+	}
+
+	return true
+}
+
+func (nw *namespaceWatcher) IsSharedGlobalService(obj interface {
+	GetAnnotations() map[string]string
+	GetNamespace() string
+}) bool {
+	// Service must be global first
+	if !nw.IsGlobalService(obj) {
+		return false
+	}
+
+	// Then check if it's also shared
+	return annotation.GetAnnotationShared(obj)
+}
+
+func (nw *namespaceWatcher) processNamespaceChange(processors []NamespaceProcessor, namespace string, isGlobal bool) {
+
+	// Normal operation: process asynchronously
+	for _, processor := range processors {
+		go processor.ProcessNamespaceChange(namespace, isGlobal)
+	}
+
+}
+
+func (nw *namespaceWatcher) updateNamespace(ns *slim_corev1.Namespace) {
+	// Check if namespace has the global annotation
+	annotationValue, hasAnnotation := annotation.Get(ns, annotation.GlobalNamespace)
+
+	nw.mu.Lock()
+
+	// Capture the previous state by querying the resource store
+	wasAnnotated := nw.isNamespaceAnnotated(ns.Name)
+	wasGlobal := nw.isNamespaceGlobalByAnnotation(ns.Name)
+	wasFilteringActive := nw.isFilteringActive()
```

Comment:
This does not work as intended, because the store already contains the up-to-date state at this point.